src/ngraph/runtime/gpu/cuda_emitter.cpp

```diff
@@ -27,3 +31,148 @@ CUDAEmitter::CUDAEmitter(GPUPrimitiveEmitter* emitter)
     : m_primitive_emitter(emitter)
 {
 }
+
+size_t CUDAEmitter::build_pad(const GPURuntimeContext* ctx,
+                              const std::array<std::string, 2>& dtypes,
+                              const ngraph::Shape& input_shape,
+                              const ngraph::Shape& output_shape,
+                              const ngraph::Shape& padding_below,
+                              const ngraph::Shape& padding_above,
+                              const ngraph::Shape& padding_interior)
+{
+    if (padding_interior.size())
+    {
+        throw std::runtime_error("Interior padding is not yet supported in the GPU transformer.");
+    }
+    std::string hash = "pad_i" + join(input_shape) + "_pb" + join(padding_below) + "_pa" +
+                       join(padding_above) + "_pi" + join(padding_interior);
+    std::replace(hash.begin(), hash.end(), ' ', '_');
+    std::replace(hash.begin(), hash.end(), ',', '_');
+
+    // For backwards compatability we currently use two unordered maps
+    // 1. one looks up the compiled cuda kernel (CudaFunctionPool)
+    // 2. the other looks to see if this kernel is already in the primitive list
+    // Once all previously implemented cuda kernels are refactored to use the
+    // CUDAEmitter/GPUPrimittiveEmitter interface, only one map (from hash to primitive index)
+    // will be required.
+
+    // check if the requested kernel is already an inserted primitive
+    size_t primitive_index = m_primitive_emitter->lookup(hash);
+    if (primitive_index != std::numeric_limits<size_t>::max())
+    {
+        return primitive_index;
+    }
+
+    // if the kernel has not been compiled, build it
+    auto compiled_kernel = ctx->compiled_kernel_pool->get(hash);
+    auto nthreads = shape_size(input_shape);
+    if (compiled_kernel == nullptr)
+    {
+        // normalize pad dimensions to shape dimensions
+        Shape pad_below(input_shape.size(), 0);
+        Shape pad_above(input_shape.size(), 0);
+        if (padding_below.size() != input_shape.size())
```

Comment:
There are two different convention for padding in nGraph, one for op::Pad and one for op::MaxPool/op::Convolution.

For op::MaxPool/Convolution, the padding rank and the tensor rank will not be equal. In that case padding_rank == tensor_rank-2, or the number of spatial dimensions. That is what this conditional is checking for. For op::Pad, you are right, the ranks will be equal.

This relates to the discussion in #824. I found this to be a little confusing, I'm not sure if the difference in convention is necessary or not. I'm guessing it is there to prevent people from accidentally padding the N and C dimensions for pooling and convolution.

Reply:
Yes, got it.